src/OrchardCore.Modules/OrchardCore.Layers/Migrations.cs

```diff
@@ -64,18 +60,25 @@ public async Task<int> UpdateFrom2Async()
                 layer.LayerRule = new Rule();
                 _conditionIdGenerator.GenerateUniqueId(layer.LayerRule);
 
-                #pragma warning disable 0618
+#pragma warning disable 0618
                 _ruleMigrator.Migrate(layer.Rule, layer.LayerRule);
 
                 layer.Rule = String.Empty;
-                #pragma warning restore 0618
+#pragma warning restore 0618
             }
 
             await _layerService.UpdateAsync(layers);
-            
-            var layerFeature = _typeFeatureProvider.GetFeatureForDependency(GetType());
-            
-            await _shellFeaturesManager.EnableFeaturesAsync(new[] { layerFeature }, force: true);
+
+            // Registered as a deferred task so the migration can complete before the shell reactivates the module causing migrations to run circularly and fail.                 
+            ShellScope.RegisterBeforeDispose((scope) =>
+            {
+                var typeFeatureProvider = scope.ServiceProvider.GetRequiredService<ITypeFeatureProvider>();
+                var layerFeature = typeFeatureProvider.GetFeatureForDependency(GetType());
+
+                // Reenable the layer feature so the new dependency of OrchardCore.Rules is activated.
+                var shellFeaturesManager = scope.ServiceProvider.GetRequiredService<IShellFeaturesManager>();
+                return shellFeaturesManager.EnableFeaturesAsync(new[] { layerFeature }, force: true);
```

Comment:
@jtkech this was the slightly problematic one.

On an RC2 site `UpdateFrom1()` has not run.
This is `UpdateFrom2()`.

When `UpdateFrom2()` starts running, it activates the feature, which starts a feedback loop, back into the data migrations, and tries to execute `UpdateFrom1()` again. Which fails, because the indexes have already been created.

Here, I register the feature reactivation as deferred, so that all migrations have run, and the migration record comitted to the db, so the next shell that starts, is reading from the latest migration record, and doesn't try to rerun the migration.

Quickly done, might be better ways to achieve it.

Reply:
@deanmarcussen Okay, LGTM

For some syncing reasons `UpdateShellDescriptorAsync()` does a `CommitAsync()`, so before the end of migrations, so anyway i prefer to defer it and if possible in its own scope. So, maybe better to use `ShellScope.AddDeferredTask()`, it will do the same but in its own scope.

Reply:
`AddDeferredTask` done. Probably what I was looking for anyway.

Reply:
Does it work? I assume yes ;)

Forgot to say that maybe better to define before `var type = GetType()` and use it in the delegate, this to prevent a closure holding a reference on the migrations scoped service, that will be used later in the context of another scope.

We could also move up both line 75 and 76, hmm but you would need to re-inject `ITypeFeatureProvider`

            var layerFeature = _typeFeatureProvider.GetFeatureForDependency(GetType());

            ShellScope.AddDeferredTask((scope) =>
            {
                // Reenable the layer feature so the new dependency of OrchardCore.Rules is activated.
                var shellFeaturesManager = scope.ServiceProvider.GetRequiredService<IShellFeaturesManager>();
                return shellFeaturesManager.EnableFeaturesAsync(new[] { layerFeature }, force: true);
            });